statsmodels/tsa/statespace/kalman_filter.py

```diff
@@ -0,0 +1,830 @@
+                if name == 'obs':
+                    continue
+                if representation[name].shape[-1] == 1:
+                    if name in kwargs:
+                        warn(warning % (name, name))
+                elif name not in kwargs:
+                    raise ValueError(exception % name)
+                else:
+                    mat = np.asarray(kwargs[name])
+                    if len(shape) == 2:
+                        validate_vector_shape('obs_intercept', mat.shape,
+                                              shape[0], nforecast)
+                        if mat.ndim < 2 or not mat.shape[1] == nforecast:
+                            raise ValueError(exception % name)
+                        representation[name] = np.c_[representation[name], mat]
+                    else:
+                        validate_matrix_shape(name, mat.shape, shape[0],
+                                              shape[1], nforecast)
+                        if mat.ndim < 3 or not mat.shape[2] == nforecast:
+                            raise ValueError(exception % name)
+                        representation[name] = np.c_[representation[name], mat]
+
+        # Construct the predicted state and covariance matrix for each time
+        # period depending on whether that time period corresponds to
+        # one-step-ahead prediction, dynamic prediction, or out-of-sample
+        # forecasting.
+
+        # If we only have simple prediction, then we can use the already saved
+        # Kalman filter output
+        if ndynamic == 0 and nforecast == 0:
+            result = self
+        else:
+            # Construct the new endogenous array.
+            endog = np.empty((self.k_endog, ndynamic + nforecast))
+            endog.fill(np.nan)
+            endog = np.asfortranarray(np.c_[self.endog[:, :nstatic], endog])
+
+            # Setup the new statespace representation
+            model_kwargs = {
+                'filter_method': self.filter_method,
+                'inversion_method': self.inversion_method,
+                'stability_method': self.stability_method,
+                'conserve_memory': self.conserve_memory,
+                'tolerance': self.tolerance,
+                'loglikelihood_burn': self.loglikelihood_burn
+            }
+            model_kwargs.update(representation)
+            model = KalmanFilter(
+                endog, self.k_states, self.k_posdef, **model_kwargs
+            )
+            model.initialize_known(
+                self.initial_state,
+                self.initial_state_cov
+            )
+            model._initialize_filter(*args, **kwargs)
+            model._initialize_state(*args, **kwargs)
+
+            result = self._predict(nstatic, ndynamic, nforecast, model)
+
+        if full_results:
+            return result
+        else:
+            return result.forecasts[:, start:end]
+
+    def _predict(self, nstatic, ndynamic, nforecast, model, *args, **kwargs):
```

Comment:
I'm just going to add a TODO note to this method, since I don't right now know where the best place to put it is.